worlds/osrs/__init__.py

```diff
@@ -0,0 +1,616 @@
+        ))
+        add_rule(self.multiworld.get_location(LocationNames.QP_X_Marks_the_Spot, self.player), lambda state: (
+            self.multiworld.get_location(LocationNames.Q_X_Marks_the_Spot, self.player).can_reach(state)
+        ))
+        add_rule(self.multiworld.get_location(LocationNames.QP_Below_Ice_Mountain, self.player), lambda state: (
+            self.multiworld.get_location(LocationNames.Q_Below_Ice_Mountain, self.player).can_reach(state)
+        ))
+
+        # place "Victory" at "Dragon Slayer" and set collection as win condition
+        self.multiworld.get_location(LocationNames.Q_Dragon_Slayer, self.player) \
+            .place_locked_item(self.create_event("Victory"))
+
+        for location_name, location in self.location_name_to_data.items():
+            location_row = self.location_rows_by_name[location_name]
+            # Set up requirements for region
+            for region_required_name in location_row.regions:
+                region_required = self.region_name_to_data[region_required_name]
+                add_rule(location, lambda state, region_required=region_required: state.can_reach(region_required, None, self.player))
+            for skill_req in location_row.skills:
+                add_rule(location, lambda state, skill_req=skill_req: self.can_reach_skill(state, skill_req.skill, skill_req.level))
+            for item_req in location_row.items:
+                add_rule(location, lambda state, item_req=item_req: state.has(item_req, self.player))
+            if location_row.qp != 0:
+                add_rule(location, lambda state, location_row=location_row: self.quest_points(state) > location_row.qp)
+        self.multiworld.completion_condition[self.player] = lambda state: (state.has("Victory", self.player))
+
+    def create_region(self, name: str) -> "Region":
+        region = Region(name, self.player, self.multiworld)
+        self.region_name_to_data[name] = region
+        self.multiworld.regions.append(region)
+        return region
+
+    def create_item(self, item_name: str) -> "Item":
+        item = [item for item in self.item_rows if item.name == item_name][0]
+        index = self.item_rows.index(item)
+        return OSRSItem(item.name, item.progression, self.base_id + index, self.player)
+
+    def create_event(self, event: str):
+        # while we are at it, we can also add a helper to create events
+        return OSRSItem(event, ItemClassification.progression, None, self.player)
+
+    def quest_points(self, state):
+        qp = 0
+        for qp_event in QP_Items:
+            if state.has(qp_event, self.player):
+                qp += int(qp_event[0])
+        return qp
+
+    """
+    Ensures a target level can be reached with available resources
+    """
+
+    def can_reach_skill(self, state, skill, level):
+        if skill == "Fishing":
+            can_train = state.can_reach(RegionNames.Shrimp, None, self.player)
+            if not self.allow_brutal_grinds:
```

Comment:
Option checking and conditionals in access rules are slow.

Instead, what you could do is make `can_reach_skill` **return a lambda** instead of a bool.

example:

```
def can_reach_skill(self, state, skill, level) -> CollectionRule:
    if skill == "Fishing":
        if self.allow_brutal_grinds or level < 5:
            return lambda state: state.can_reach(RegionNames.Shrimp, None, self.player)
        if level < 20:
            return lambda state: (state.can_reach(RegionNames.Shrimp, None, self.player)
                                  and state.can_reach(RegionNames.Port_Sarim, None, self.player))
        return lambda state: (state.can_reach(RegionNames.Shrimp, None, self.player) 
                              and state.can_reach(RegionNames.Fly_Fish, None, self.player))
```

I know it looks terrible, but this means that the access rule itself can just instantly evaluate every time.
Right now, **every time the location is checked**, it needs to check the option, check the level, etc. That's very slow. The majority of generation time is going to be spent checking location and entrance access rules, so you should make sure they evaluate fast. That should be your priority.

To make a stronger statement: Option checking *inside* access rules is disallowed, as far as I and some other prominent/core devs are concerned.

Reply:
If you take anything important away from my review, it's this. This absolutely needs to be done differently. You have to set different rules based on options / parameters, rather than checking for options / parameters inside the rules.